### PR TITLE
Update cpp_interface.dd for recent developments

### PR DESCRIPTION
**Pull request by JakobOvrum**
* Removes old assumptions
 * Specifies the new type and template stuff, with examples
 * Updated memory allocation section
 * Updated data type table

Depends on: [issue 15372](https://issues.dlang.org/show_bug.cgi?id=15372). The template example doesn't link.

The first commit contains only the tabs to spaces changes, so it's probably easier to review using the diff for the second commit.

### CODE REVIEW

File: cpp_interface.dd [hunk 1]
```diff
@@ -296,328 +340,406 @@ class E
...
+    $(ARGS $(B unsigned char))
+    )
+
+    $(TROW
+    $(ARGS $(B char)),
+    $(ARGS $(B char) (chars are unsigned in D))
+    )
+
+    $(TROW
+    $(ARGS $(B core.stdc.stddef.wchar_t)),
```
Comment (JackStouffer) [inline cpp_interface.dd]: This line seems to be causing some error in the LaTex generation.

Comment (andralex) [inline cpp_interface.dd]: It's because of the underscore I think. Use `$(UNDERSCORE)` instead.

Comment (CyberShadow) [inline cpp_interface.dd]: Or use `$(D ...)`, as is more appropriate here because you're referring to an identifier.

Comment (JakobOvrum) [inline cpp_interface.dd]: Yeah, that makes sense. I changed it to use `$(D ...)` and the test passes now.

**Comment by JakobOvrum** [comment]
I filed the linking issue: https://issues.dlang.org/show_bug.cgi?id=15372

**Comment by JackStouffer** [comment]
So this is blocked by 15372? If so it would be useful if someone added the "Blocked" tag to this.

**Comment by JakobOvrum** [comment]
Well, I could just remove function templates from the article. I just can't claim it works when it doesn't.

**Comment by JackStouffer** [comment]
I'd say add a note about bug 15372 and then merge this, as this is how it should work baring bugs and that's how we document everything else.

**Comment by JackStouffer** [comment]
@JakobOvrum rebase please

**Comment by JakobOvrum** [comment]
Is the `BUGZILLA` macro broken for latex output?

**Comment by CyberShadow** [comment]
> Is the BUGZILLA macro broken for latex output?

Seems so.

**Comment by deadalnix** [comment]
Is there a way to see the page somewhere ?

**Comment by JakobOvrum** [comment]
This PR now depends on #1194 for `MREF`.

> Is there a way to see the page somewhere ?

The build is failing because of a missing LaTeX macro (added in #1194), so I guess not.

**Comment by adamdruppe** [comment]
I just merged 1194, if the test re-runs here, ping me and I'll merge this too.

**Comment by tsbockman** [comment]
@adamdruppe The test has passed now.

**Comment by tsbockman** [comment]
Thanks for this. When will this actually go live on the website?

**Comment by CyberShadow** [comment]
The next time I, Martin or Andrei syncs it.

Would you like me to sync it now?

**Comment by tsbockman** [comment]
@CyberShadow 
> Would you like me to sync it now?

That would be nice. Thanks.

People ask about C++ support from time to time on the forums, and I'd like to finally be able to just say, "go read the spec".

**Comment by CyberShadow** [comment]
Done: http://dlang.org/spec/cpp_interface.html

**Comment by tsbockman** [comment]
Thanks. Any idea why "*COM interface*" at the top of the page is a dead link?

**Comment by CyberShadow** [comment]
It seems to be using the wrong macro. But http://dlang.org/COM.html is just a redirect to a wiki page with one line in it. The link should probably just be removed.

**Comment by tsbockman** [comment]
@CyberShadow 
> The link should probably just be removed.

I made a PR: https://github.com/D-Programming-Language/dlang.org/pull/1249

**Pull request merged**